File: ruby/bin/shop.rb
#! /usr/bin/env ruby

require_relative '../lib/basket/basket'
require_relative '../lib/basket_actions/basket_actions'
require_relative '../lib/catalogue_actions/catalogue_actions'
require_relative '../lib/filestore/catalogue_reader'
require_relative '../lib/filestore/warehouse_reader'
require_relative '../lib/products/catalogue'
require_relative '../lib/user_interface/user_interface'
require_relative '../lib/warehouse/warehouse'
require_relative '../lib/warehouse_actions/warehouse_actions'

class Application

	class << self

		def main(args)
			displayWelcomeMessage
			warehouse = Warehouse.fromFile(WarehouseReader.new("../warehouse.dat"))
			catalogue = Catalogue.fromFile(CatalogueReader.new("../catalogue.dat"))
			basket = Basket.new
			catalogueActions = CatalogueActions.new(catalogue)
			warehouseActions = WarehouseActions.new(warehouse, catalogue)
			basketActions = BasketActions.new(basket, catalogue, warehouse)
			UserInterface.new(System.in, catalogueActions, warehouseActions, basketActions).start
			displayGoodbyeMessage
			exit 0
		end

		private

		def displayGoodbyeMessage
			puts
			puts "Goodbye. Thanks for your custom!"
			puts
		end

		def displayWelcomeMessage
			puts "Welcome to our little corner shop!"
			puts
			puts "For help, type 'h' or 'help' or '?'"
			puts
		end
	end

end

Application.main $*

File: python/shop.py
#! /usr/bin/env python

import sys
from basket.basket import Basket
from basket_actions.basket_actions import BasketActions
from catalogue_actions.catalogue_actions import CatalogueActions
from filestore.catalogue_reader import CatalogueReader
from filestore.warehouse_reader import WarehouseReader
from products.catalogue import Catalogue
from user_interface.user_interface import UserInterface
from warehouse.warehouse import Warehouse
from warehouse_actions.warehouse_actions import WarehouseActions

class Application:

    def main(self):
        self.displayWelcomeMessage()
        warehouse = Warehouse.fromFile(WarehouseReader("../warehouse.dat"))
        catalogue = Catalogue.from_file(CatalogueReader("../catalogue.dat"))
        basket = Basket()
        catalogueActions = CatalogueActions(catalogue)
        warehouseActions = WarehouseActions(warehouse, catalogue)
        basketActions = BasketActions(basket, catalogue, warehouse)
        UserInterface(catalogueActions, warehouseActions, basketActions).start()
        self.displayGoodbyeMessage()
        sys.exit()

    def displayGoodbyeMessage(self):
        print
        print "Goodbye. Thanks for your custom!"
        print

    def displayWelcomeMessage(self):
        print "Welcome to our little corner shop!"
        print
        print "For help, type 'h' or 'help' or '?'"
        print

Application().main()

